Source/Request.swift

```diff
@@ -38,17 +38,26 @@ public class Request {
     /// - cancelled: Set when `cancel()` is called. Any tasks created for the `Request` will have `cancel()` called on
     ///              them. Unlike `resumed` or `suspended`, once in the `cancelled` state, the `Request` can no longer
     ///              transition to any other state.
+    /// - finished:  Set when all response serialization completion closures have been cleared on the `Request` and
+    ///              queued on their respective queues.
     public enum State {
-        case initialized, resumed, suspended, cancelled
+        case initialized, resumed, suspended, cancelled, finished
 
         /// Determines whether `self` can be transitioned to `state`.
         func canTransitionTo(_ state: State) -> Bool {
             switch (self, state) {
-            case (.initialized, _): return true
-            case (_, .initialized), (.cancelled, _): return false
-            case (.resumed, .cancelled), (.suspended, .cancelled),
-                 (.resumed, .suspended), (.suspended, .resumed): return true
-            case (.suspended, .suspended), (.resumed, .resumed): return false
+            case (.initialized, _):
```

Comment:
I think the state machine need to be updated to ensure (or make explicit) we can only transition to `finished` from non-`cancelled` states.

Reply:
I agree. I actually put that code in and it's covered by this case already:

```swift
case (_, .initialized), (.cancelled, _):
    return false
```

Reply:
To make it more clear, I updated that case to this:

```swift
case (_, .initialized), (.cancelled, _), (.finished, _):
    return false
```